# Exploring eBay Car Sales Data

Project goals:
- clean the data,
- analyze the included used car listings.

In these project we'll be working with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data). For our project purposed we will be working a sample of 50 000 data points from full dataset.

## Data import

Let's start by importing the libraries we need and reading the dataset into pandas.

In [1]:
import pandas as pd

In [2]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

Let's render the first few and last few values of any pandas object.

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From abowe information we see that this dataset has 50,000 entries and 20 columns. There are five columns with int64 data types and the rest are boject type.  Columns `vehicleType`, `gearbox`, `model`, `fuelType` and `notRepairedDamage` have some null values.

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Some columns like price and odometer can be represented numerically. We will need to be cleaned to fit this data type.

In below table we have information about data in columns.

| Column Name | Description |
| ----------- | ----------- |
| dateCrawled | When this ad was first crawled. All field-values are taken from this date |
| name | Name of the car |
| seller | Whether the seller is private or a dealer |
| offerType | The type of listing |
| price | The price on the ad to sell the car |
| abtest | Whether the listing is included in an A/B test |
| vehicleType | The vehicle Type |
| yearOfRegistration | The year in which the car was first registered |
| gearbox | The transmission type |
| powerPS | The power of the car in PS |
| model | The car model name |
| odometer | How many kilometers the car has driven |
| monthOfRegistration | The month in which the car was first registered |
| fuelType | What type of fuel the car uses |
| brand | The brand of the car |
| notRepairedDamage | If the car has a damage which is not yet repaired |
| dateCreated | The date on which the eBay listing was created |
| nrOfPictures | The number of pictures in the ad |
| postalCode | The postal code for the location of the vehicle |
| lastSeenOnline | When the crawler saw this ad last online |

### Cleaning Column Names

We will print an array of the existing column names.

In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

As we can see columns are in camels style while for easier programming in python preferable style is snakecase. Below all the column labels was converted to snakecase and some labels was shortened where necessary by using `DataFrame.columns` attribute.

In [7]:
autos.columns = ["date_crawled", "name", "seller", "offer_type", "price", "ab_test",
       "vehicle_type", "registration_year", "gear_box", "power_ps", "model",
       "odometer", "registration_month", "fuel_type", "brand",
       "unrepaired_damage", "ad_created", "num_photos", "postal_code",
       "last_seen"]

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

### Initial Exploration and Cleaning.

Let's do some basic data exploration to determine what more cleaning task need to be done. we will be checking for:
- Text columns where almost all values are the same. The may not provide any useful information during analysis so they can be dropped.
- Columns with numeric data stored as text. This needs to be cleaned and converted.

`DataFrame.describe()` method with `include="all"` will give the descriptive statistics for all numeric and categorical columns for closer look at it.

In [9]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-27 22:55:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Based on the provided data, here are some more observations:

**Columns that have mostly one value and can be dropped:**
- `"num_photos"`: column has zero pictures which warrants it to be dropped.
- `"seller"`: contains only two unique values and has mostly one value "privat" (49999 out of 50000) which is not useful in our analysis.
- `"offer_type"`: contains only two unique values and has mostly one value "Angebot" (49999 out of 50000) which is also not useful in our analysis.
- `"ab_test"`: contains only two unique values and more than halfe of data is  one value "test" (25756 out of 50000), also not usefull for our analysis.
- `"gear_box"`: contains mostly one value "manuell" (36993 out of 50000) and has only two unique values but it can be usefull for us.
- `"unrepaired_damage"`: contains mostly one value "nein" (35232 out of 50000) and has only two unique values which also can be usefull for us.

In [10]:
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [11]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [12]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [13]:
autos["ab_test"].value_counts()

test       25756
control    24244
Name: ab_test, dtype: int64

In [14]:
autos["gear_box"].value_counts()

manuell      36993
automatik    10327
Name: gear_box, dtype: int64

In [15]:
autos["unrepaired_damage"].value_counts()

nein    35232
ja       4939
Name: unrepaired_damage, dtype: int64

Now we will remove abowe colums because they are not usefull for our analysis

In [16]:
autos = autos.drop(["seller", "offer_type", "num_photos", "ab_test"], axis=1)

**Colums which are examples of numeric data stored as text that needs to be cleaned:**
- `"odometer"`: contains value which is numeric data with information about SI unit of length "km", so we need to chane the name of the coulmn and convert data to only contains numeric value.
- `"pice"`: contains value which is numeric data with information about the currency, so also here we need to remove text and convert column to numeric with different column name

### Exploring the Odometer and Price Columns

In [17]:
autos["odometer"].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [18]:
autos["price"].value_counts()

$0          1421
$500         781
$1,500       734
$2,500       643
$1,200       639
            ... 
$10,790        1
$6,447         1
$679           1
$29,970        1
$114,400       1
Name: price, Length: 2357, dtype: int64

### Removing non-numeric characters and rename in `price` and `odometer` columns

Now we will:
- remove any non-numeric characters and convert the column to a numeric type
- rename the `odometer` column to `odometer_km` and the `price` column to `price_usd` for specificity

In [19]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

In [20]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)

In [21]:
autos.rename({"price": "price_usd"}, axis=1, inplace=True)

In [22]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_usd vehicle_type  registration_year   gear_box  power_ps   model  \
0       5000          bus               2004    manuell       158  andere   
1       8500    limousine               1997  automatik       286     7er   
2       8990    limousine               2009    manuell       102    golf   
3       4350   kleinwagen               2007  automatik        71  fortwo   
4       1350        kombi               2003    manuell         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot              nein   
1       150000                   6    benzin         bmw              nein   
2        70000                   7    benzin  volkswagen              nein   
3        70000                   6    benzin       smart              nein   
4       150000                   7    benzin        ford              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Continue exploring `odometer_km` and `price_usd` colums to specifically looking for data that doesn't look right.

##### `odometer_km` column checking

Let's check how many unique values are in `odometer_km` column

In [23]:
autos["odometer_km"].unique().shape

(13,)

In `odometer_km` column there are 13 unique values

`Series.describe()` let us view min/max/median/mean etc.

In [24]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

`Series.value_counts()` return a Series containing counts of unique values. We can combain it with variation:
- `.head()`: to check only first five values
- `Series.sort_index()` with `ascending= True` or `False` let us view the highest and lowest values with their counts (can also chain to `head()` here).

In [25]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

In [26]:
autos["odometer_km"].value_counts().sort_index(ascending=False).head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [27]:
autos["odometer_km"].value_counts().sort_index(ascending=True).head()

5000     967
10000    264
20000    784
30000    789
40000    819
Name: odometer_km, dtype: int64

The `odometer_km` column is rounded with milage ranges from listings within 5000 to 150,000 kilometers. Over half of the listings 32,424 fall in the 150,000 range. This may indicate that the dataset includes older cars or cars that have been driven extensively.

There are very few cars in the dataset with low odometer readings - below 50,000. These colud suggest that the dataset is not representative of the market for relatively new or low-mileage used cars.


##### `price_usd` column checking

Now we will furhter investigate the `price_usd` column

In [28]:
autos["price_usd"].unique().shape

(2357,)

In `price_usd` column there are 2357 unique values.

In [29]:
autos["price_usd"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [30]:
autos["price_usd"].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price_usd, Length: 2357, dtype: int64

In [31]:
autos["price_usd"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_usd, dtype: int64

In [32]:
autos["price_usd"].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_usd, dtype: int64

As we can see there are some prices that are very unlikely. It is very unlikely that the price of a car could be less than 200 usd. In addition, it is unlikely that a second hand car can cost more then 1 mln USD. Whats more The price values increase steadly upto 350,000 then jump drastically to 1 million. 

Let's check how the data will be look when we removed values more than 350,000 and less than 200.

In [33]:
autos = autos[autos["price_usd"].between(200,350000)]

In [34]:
autos["price_usd"].describe()

count     47645.000000
mean       6000.707273
std        9110.783444
min         200.000000
25%        1300.000000
50%        3190.000000
75%        7500.000000
max      350000.000000
Name: price_usd, dtype: float64

In [35]:
autos["price_usd"].value_counts().sort_index().head(20)

200    266
205      1
210      1
215      2
217      1
219      1
220     33
222     12
225      8
230     12
235      2
238      1
240      3
248      1
249     13
250    291
251      1
255      1
260      5
269      1
Name: price_usd, dtype: int64

Now data looks much better but still not so good. Let ch.eck another price range

In [36]:
autos = autos[autos["price_usd"].between(200,200001)]
autos["price_usd"].describe()

count     47637.000000
mean       5953.790079
std        8342.836867
min         200.000000
25%        1300.000000
50%        3180.000000
75%        7500.000000
max      198000.000000
Name: price_usd, dtype: float64

Now the mean price is now 5,953, which is a more reasonable value than the mean price before removing outliers. The standard deviation has also decreased, indicating that the data is more concentrated around the mean. Additionally, the minimum price is now 200, which is the lower limit we set for the price range, and the maximum price is 200,000, which is a more reasonable value.

**Columns that need more investigation:***
- `"vehicle_type"`: missing values (5000 out of 50000) and contains multiple categories that need further investigation.
- `"registration_year"`: the minimum value is 1000 and the maximum value is 9999, which is unrealistic and needs further investigation.
- `"model"`: contains missing values (2758 out of 50000) and many different categories that need further investigation.
- `"registration_month"`: missing values (5000 out of 50000) and the need for further investigation.
- `"fuel_type"`: missing values (4482 out of 50000) and multiple categories that need further investigation.
- `"ad_created"`: contains not so many unique values (76 out of 50000), so it's difficult to draw any conclusions without more information on what this column represents.
- `"data_crawled"`: contains many unique values (48213 of 50000), it shows information aboud data and time when was first crawled, need for further investigation.
- `"last_seen"`: contains many unique values (39481 of 50000), it shows information aboud data and time the crawler saw this ad last online, need for further investigation.

### Exploring the date columns

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

In [37]:
autos["date_crawled"].describe()

count                   47637
unique                  46022
top       2016-03-21 20:37:19
freq                        3
Name: date_crawled, dtype: object

In [38]:
autos["last_seen"].describe()

count                   47637
unique                  37804
top       2016-04-07 06:17:27
freq                        8
Name: last_seen, dtype: object

In [39]:
autos["ad_created"].describe()

count                   47637
unique                     76
top       2016-04-03 00:00:00
freq                     1856
Name: ad_created, dtype: object

In [40]:
autos["registration_month"].describe()

count    47637.000000
mean         5.822638
std          3.667247
min          0.000000
25%          3.000000
50%          6.000000
75%          9.000000
max         12.000000
Name: registration_month, dtype: float64

In [41]:
autos["registration_month"].unique()

array([ 3,  6,  7,  4,  8, 12, 10,  0,  9, 11,  5,  2,  1])

As we can see in our dataset in column `registration_month` we have value `0`. In calenadar we have only 12 months so that can mean that the value `0` refers to an unknown month of registration.

In [42]:
autos["registration_year"].describe()

count    47637.000000
mean      2004.799631
std         88.431141
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

- The data consists of 47,637 records of registration year for vehicles.
- The mean registration year is 2004.79.
- The standard deviation of registration year is 88.64.
- The minimum value for the registration_year is 1000, before cars were invented. This suggests that the value is likely to be incorrect, possibly due to a data entry error or a mistake in the dataset.
- The 25th percentile of registration year is 1999, meaning that 25% of the vehicles were registered before 1999.
- The median (50th percentile) registration year is 2004, meaning that half of the vehicles were registered before 2004 and half were registered after.
- The 75th percentile of registration year is 2008, meaning that 75% of the vehicles were registered before 2008.
- The maximum value for the registration_year is 9999, many years into the future. This suggests that the value is also likely to be incorrect, possibly due to a data entry error or a mistake in the dataset.

As we can see `date_crawled`, `ad_created` and `last_seen` columns are identified as string values. The other two columns: `registration_month` and `registration_year` are represented as numeric values.

Let's get closer look at `date_crawled`, `ad_created` and `last_seen` column.

In [43]:
autos[["date_crawled","ad_created","last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Observe that the first 10 character in the srting represent dates in the format year-month-day. I used this observation to explore the columns furhter.

In [44]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036001
2016-03-08    0.033126
2016-03-09    0.033042
2016-03-10    0.032328
2016-03-11    0.032706
2016-03-12    0.036883
2016-03-13    0.015702
2016-03-14    0.036568
2016-03-15    0.034238
2016-03-16    0.029452
2016-03-17    0.031551
2016-03-18    0.012826
2016-03-19    0.034616
2016-03-20    0.037807
2016-03-21    0.037345
2016-03-22    0.032706
2016-03-23    0.032370
2016-03-24    0.029326
2016-03-25    0.031425
2016-03-26    0.032223
2016-03-27    0.031194
2016-03-28    0.034994
2016-03-29    0.033986
2016-03-30    0.033881
2016-03-31    0.031845
2016-04-01    0.033818
2016-04-02    0.035687
2016-04-03    0.038709
2016-04-04    0.036568
2016-04-05    0.013120
2016-04-06    0.003170
2016-04-07    0.001364
Name: date_crawled, dtype: float64

From the above results, we can observe that ads in the dataset were crawled between 2016-03-05 and 2016-04-07. It appears that the indexing rate generally remained constant and then dropped sharply around 2016-04-06. We can also see some drops in the indexing rate (for example: 2016-03-06 and 2016-03-13), which may indicate a lower level of online activity during these periods.

In [45]:
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.001092
2016-03-06    0.004303
2016-03-07    0.005374
2016-03-08    0.007179
2016-03-09    0.009614
2016-03-10    0.010475
2016-03-11    0.012301
2016-03-12    0.023910
2016-03-13    0.008901
2016-03-14    0.012532
2016-03-15    0.015723
2016-03-16    0.016290
2016-03-17    0.028087
2016-03-18    0.007284
2016-03-19    0.015618
2016-03-20    0.020656
2016-03-21    0.020509
2016-03-22    0.021412
2016-03-23    0.018494
2016-03-24    0.019628
2016-03-25    0.019082
2016-03-26    0.016689
2016-03-27    0.015492
2016-03-28    0.020740
2016-03-29    0.022189
2016-03-30    0.024582
2016-03-31    0.023868
2016-04-01    0.022902
2016-04-02    0.024813
2016-04-03    0.025086
2016-04-04    0.024624
2016-04-05    0.125449
2016-04-06    0.222432
2016-04-07    0.132670
Name: last_seen, dtype: float64

We can observe that most of the ads were last seen between 2016-04-05 and 2016-04-07. Additionally, the values steadily increase from 2016-03-05 to 2016-04-04 and then suddenly jump on 5th og April. This suggests that there was an overall increase in the number of listings during this period, which may indicate a rise in car sales or a greater number of cars being added to the website.

In [46]:
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038961
2016-04-04    0.036925
2016-04-05    0.011840
2016-04-06    0.003254
2016-04-07    0.001218
Name: ad_created, Length: 76, dtype: float64

The ads creation dates contains many entries. The earliest advertisement was created on 2015-06-11. It seems that the frequency of ad creation remained relatively low and stable for most of the period1. However, there was a significant spike in ad creation frequency on 2016-03-05 (0.022944), which may indicate a sudden increase in advertising activity.

#### Dealing with Incorrect Registration Year Data

Car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. The car was invented in 1886. So for our analysis we will take interval from 1886 to 2016.

It is safe to remove the rows that fall outside this interval as they are likely to be inaccurate or erroneous data points.

In [47]:
autos = autos[autos["registration_year"].between(1886,2016)]

In [48]:
autos["registration_year"].value_counts(normalize=True)

2000    0.065949
2005    0.063480
1999    0.062148
2004    0.058849
2003    0.058806
          ...   
1938    0.000022
1939    0.000022
1953    0.000022
1943    0.000022
1952    0.000022
Name: registration_year, Length: 78, dtype: float64

In [49]:
autos["registration_year"].value_counts(normalize=True).sort_index().head(20)

1910    0.000044
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000044
1937    0.000087
1938    0.000022
1939    0.000022
1941    0.000044
1943    0.000022
1948    0.000022
1950    0.000022
1951    0.000044
1952    0.000022
1953    0.000022
1954    0.000044
1955    0.000044
1956    0.000087
1957    0.000044
1958    0.000087
Name: registration_year, dtype: float64

In [50]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False).head(25)

2016    0.025318
2015    0.008170
2014    0.014330
2013    0.017410
2012    0.028529
2011    0.035323
2010    0.034667
2009    0.045415
2008    0.048167
2007    0.049631
2006    0.058259
2005    0.063480
2004    0.058849
2003    0.058806
2002    0.053956
2001    0.057014
2000    0.065949
1999    0.062148
1998    0.049915
1997    0.040456
1996    0.028442
1995    0.024444
1994    0.012954
1993    0.008869
1992    0.007755
Name: registration_year, dtype: float64

When the values outside the upper and lower limits are removed, we observe that most of vehicles (about 94%) were registered between 1994 and 2016. The distribution also shows a decrease in the percentage of vehicles registered in earlier years. This indicates that the majority of vehicles in the dataset are relatively new models.

### Exploring Price by Brand

In [51]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211281
bmw               0.111232
opel              0.105924
mercedes_benz     0.097645
audi              0.087619
ford              0.068745
renault           0.046551
peugeot           0.029949
fiat              0.025099
seat              0.018175
skoda             0.016514
nissan            0.015291
mazda             0.015138
smart             0.014374
citroen           0.014133
toyota            0.012932
hyundai           0.010114
sonstige_autos    0.009524
volvo             0.009197
mini              0.008891
mitsubishi        0.008170
honda             0.007930
kia               0.007143
alfa_romeo        0.006684
porsche           0.005985
suzuki            0.005811
chevrolet         0.005723
chrysler          0.003561
dacia             0.002687
daihatsu          0.002490
jeep              0.002316
land_rover        0.002141
subaru            0.002097
saab              0.001660
jaguar            0.001529
daewoo            0.001485
rover             0.001333
t

In [52]:
autos["brand"].describe()

count          45778
unique            40
top       volkswagen
freq            9672
Name: brand, dtype: object

Exploring the unique values in the `"brand"` column, we find that there are 40 unique car brands in the dataset. Some of the most common brands include Volkswagen, BMW, Opel, Mercedes-Benz, Audi, Ford etc. To simplify our analysis, we will choose to aggregate data on the top 15 brands in the dataset.

In [53]:
top_brands = autos["brand"].value_counts(normalize=True).head(15).index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen'],
      dtype='object')

We choose to aggregate data on these top 15 brands because aggregating on too many brands may result in less meaningful insights.

Next, we create an empty dictionary to hold our aggregate data:

In [54]:
brand_mean_price = {}

for brand in top_brands:
    selected_row = autos[autos["brand"] == brand]
    mean_value = selected_row["price_usd"].mean()
    brand_mean_price[brand] = round(mean_value, 2)

In [55]:
brand_mean_price

{'volkswagen': 5506.44,
 'bmw': 8306.01,
 'opel': 3077.58,
 'mercedes_benz': 8691.72,
 'audi': 9406.09,
 'ford': 3883.29,
 'renault': 2552.52,
 'peugeot': 3142.02,
 'fiat': 2925.95,
 'seat': 4505.46,
 'skoda': 6451.04,
 'nissan': 4829.11,
 'mazda': 4204.59,
 'smart': 3596.4,
 'citroen': 3818.96}

Analyzing the results we see that the top three brands with the highest mean prices are BMW, Mercedes-Benz and Audi, while  Renault, Fiat, and Opel have the lowest mean prices. This suggests that brand plays an important role in determining the price of a car in the secondary market.

### Storing Aggregate Data in a DataFrame

In [56]:
brand_mean_mileage = {}

for brand in top_brands:
    selected_row = autos[autos["brand"] == brand]
    mean_value = selected_row["odometer_km"].mean()
    brand_mean_mileage[brand] = round(mean_value, 2)

In [57]:
brand_mean_mileage

{'volkswagen': 128774.81,
 'bmw': 132839.75,
 'opel': 129231.8,
 'mercedes_benz': 131091.72,
 'audi': 129260.78,
 'ford': 124095.96,
 'renault': 128052.56,
 'peugeot': 126929.25,
 'fiat': 116949.52,
 'seat': 121604.57,
 'skoda': 110998.68,
 'nissan': 118178.57,
 'mazda': 124076.48,
 'smart': 99734.04,
 'citroen': 119814.53}

Analyzing the results we see that the top three brands with the highest mean mileage are BMW, Mercedes-Benz and Audi, while Smart, Skoda and Fiat have the lowest mean mileage.

Now we will convert dictionaries: `brand_mean_price` and `brand_mean_mileage` in to pandas series.

In [58]:
mean_mileage = pd.Series(brand_mean_mileage)
mean_price = pd.Series(brand_mean_price)

In [59]:
mean_mileage

volkswagen       128774.81
bmw              132839.75
opel             129231.80
mercedes_benz    131091.72
audi             129260.78
ford             124095.96
renault          128052.56
peugeot          126929.25
fiat             116949.52
seat             121604.57
skoda            110998.68
nissan           118178.57
mazda            124076.48
smart             99734.04
citroen          119814.53
dtype: float64

In [60]:
mean_price

volkswagen       5506.44
bmw              8306.01
opel             3077.58
mercedes_benz    8691.72
audi             9406.09
ford             3883.29
renault          2552.52
peugeot          3142.02
fiat             2925.95
seat             4505.46
skoda            6451.04
nissan           4829.11
mazda            4204.59
smart            3596.40
citroen          3818.96
dtype: float64

When aur data is saved as pandas series we can create a datafreame from these series. We will save first series object `mean_price` as dataframe and assigne to it the second sereies object `mean_mileage` as a new column.

In [61]:
brand_info = pd.DataFrame(mean_price, columns=["mean_price"])
brand_info

mean_price
volkswagen        5506.44
bmw               8306.01
opel              3077.58
mercedes_benz     8691.72
audi              9406.09
ford              3883.29
renault           2552.52
peugeot           3142.02
fiat              2925.95
seat              4505.46
skoda             6451.04
nissan            4829.11
mazda             4204.59
smart             3596.40
citroen           3818.96

In [62]:
brand_info["mean_mileage"] = mean_mileage
brand_info

mean_price  mean_mileage
volkswagen        5506.44     128774.81
bmw               8306.01     132839.75
opel              3077.58     129231.80
mercedes_benz     8691.72     131091.72
audi              9406.09     129260.78
ford              3883.29     124095.96
renault           2552.52     128052.56
peugeot           3142.02     126929.25
fiat              2925.95     116949.52
seat              4505.46     121604.57
skoda             6451.04     110998.68
nissan            4829.11     118178.57
mazda             4204.59     124076.48
smart             3596.40      99734.04
citroen           3818.96     119814.53

In [63]:
brand_info.describe()

mean_price   mean_mileage
count    15.000000      15.000000
mean   4993.145333  122775.534667
std    2224.722961    8756.772040
min    2552.520000   99734.040000
25%    3369.210000  118996.550000
50%    4204.590000  124095.960000
75%    5978.740000  129003.305000
max    9406.090000  132839.750000

* The brand with the highest mean mileage is BMW, followed by Mercedes Benz and Audi.

* The brand with the highest mean price is Audi, followed by Mercedes Benz and BMW.

* There seems to be a correlation between mean mileage and mean price, as the brands with the highest mean price also have high mean mileage.

* Renault, Fiat and Opel seem to be the more affordable brands with relatively low mean prices.

* Nissan  falls in the middle range in terms of mean price, with also mean mileage near to mean value of top 15 brands.

* 25% of top 15 brands cost less than 3369.21 USD and have less than 118996.55 km mileages.

* 50% of top 15 brands cost less than 4204.59 USD and have less than 124095.96 km mileages.

* 75% of top 15 brands cost less than 5978.74 USD and have less than 129003.305 km mileages.

* The mean price is 4993.145333 USD and the mean mileage is 122775.534667.

### Data cleaning next steps

In [64]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price_usd vehicle_type  registration_year   gear_box  power_ps   model  \
0           5000          bus               2004    manuell       158  andere   
1           8500    limousine               1997  automatik       286     7er   
2           8990    limousine               2009    manuell       102    golf   
3           4350   kleinwagen               2007  automatik        71  fortwo   
4           1350        kombi               2003    manuell         0   focus   
...          ...          ...                ...        ...       ...     ...   
49995      24900    limousine               2011  automatik       239      q5   
49996       1980       cabrio               1996    manuell        75   astra   
49997      13200       cabrio               2014  automatik        69     500   
49998      22900        kombi               2013    manuell       150      a3   
49999       1250    limousine               1996    manuell       101  vectra   

       odometer_km  registration_month fuel_type       brand  \
0           150000                   3       lpg     peugeot   
1           150000                   6    benzin         bmw   
2            70000                   7    benzin  volkswagen   
3            70000                   6    benzin       smart   
4           150000                   7    benzin        ford   
...            ...                 ...       ...         ...   
49995       100000                   1    diesel        audi   
49996       150000                   5    benzin        opel   
49997         5000                  11    benzin        fiat   
49998        40000                  11    diesel        audi   
49999       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                  nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                  nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                  nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                  nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                  nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995              nein  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996              nein  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997              nein  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998              nein  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999              nein  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[45778 rows x 16 columns]

Looking at our dataset we can find some columns where some data is in german: `vehicle_type`, `gear_box` and `unrepaired_damage`. Lets take closer look to it and change german words to their english version.

In [65]:
autos["gear_box"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [66]:
autos["unrepaired_damage"].unique()

array(['nein', nan, 'ja'], dtype=object)

In [67]:
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'kleinwagen', 'kombi', nan, 'coupe', 'suv',
       'cabrio', 'andere'], dtype=object)

In [68]:
gear = {"manuell": "manual", "automatik": "automatic"}
demage = {"ja": "yes", "nein": "no"}
types = {"bus": "bus", "limousine": "limousine", "kleinwagen": "supermini", "kombi":"kombi", "coupe":"coupe", "suv": "suv", "cabrio": "cabrio", "andere": "other"}

In [69]:
autos["gear_box"] = autos["gear_box"].map(gear)
autos["unrepaired_damage"] = autos["unrepaired_damage"].map(demage)
autos["vehicle_type"] = autos["vehicle_type"].map(types)

In [70]:
autos

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       price_usd vehicle_type  registration_year   gear_box  power_ps   model  \
0           5000          bus               2004     manual       158  andere   
1           8500    limousine               1997  automatic       286     7er   
2           8990    limousine               2009     manual       102    golf   
3           4350    supermini               2007  automatic        71  fortwo   
4           1350        kombi               2003     manual         0   focus   
...          ...          ...                ...        ...       ...     ...   
49995      24900    limousine               2011  automatic       239      q5   
49996       1980       cabrio               1996     manual        75   astra   
49997      13200       cabrio               2014  automatic        69     500   
49998      22900        kombi               2013     manual       150      a3   
49999       1250    limousine               1996     manual       101  vectra   

       odometer_km  registration_month fuel_type       brand  \
0           150000                   3       lpg     peugeot   
1           150000                   6    benzin         bmw   
2            70000                   7    benzin  volkswagen   
3            70000                   6    benzin       smart   
4           150000                   7    benzin        ford   
...            ...                 ...       ...         ...   
49995       100000                   1    diesel        audi   
49996       150000                   5    benzin        opel   
49997         5000                  11    benzin        fiat   
49998        40000                  11    diesel        audi   
49999       150000                   1    benzin        opel   

      unrepaired_damage           ad_created  postal_code            last_seen  
0                    no  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                    no  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                    no  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                    no  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                    no  2016-04-01 00:00:00        39218  2016-04-01 14:38:50  
...                 ...                  ...          ...                  ...  
49995                no  2016-03-27 00:00:00        82131  2016-04-01 13:47:40  
49996                no  2016-03-28 00:00:00        44807  2016-04-02 14:18:02  
49997                no  2016-04-02 00:00:00        73430  2016-04-04 11:47:27  
49998                no  2016-03-08 00:00:00        35683  2016-04-05 16:45:07  
49999                no  2016-03-13 00:00:00        45897  2016-04-06 21:18:48  

[45778 rows x 16 columns]

Now we will convert the dates to be uniform numeric data, so `"2016-03-21"` becomes the integer `20160321`.

In [71]:
autos["date_crawled"] = pd.to_datetime(autos["date_crawled"], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)
autos["ad_created"] = pd.to_datetime(autos["ad_created"], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)
autos["last_seen"] = pd.to_datetime(autos["last_seen"], format='%Y-%m-%d').dt.strftime('%Y%m%d').astype(int)

In [72]:
autos

date_crawled                                               name  \
0          20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   
1          20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2          20160326                         Volkswagen_Golf_1.6_United   
3          20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4          20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...             ...                                                ...   
49995      20160327   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996      20160328  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997      20160402                     Fiat_500_C_1.2_Dualogic_Lounge   
49998      20160308                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999      20160314                                Opel_Vectra_1.6_16V   

       price_usd vehicle_type  registration_year   gear_box  power_ps   model  \
0           5000          bus               2004     manual       158  andere   
1           8500    limousine               1997  automatic       286     7er   
2           8990    limousine               2009     manual       102    golf   
3           4350    supermini               2007  automatic        71  fortwo   
4           1350        kombi               2003     manual         0   focus   
...          ...          ...                ...        ...       ...     ...   
49995      24900    limousine               2011  automatic       239      q5   
49996       1980       cabrio               1996     manual        75   astra   
49997      13200       cabrio               2014  automatic        69     500   
49998      22900        kombi               2013     manual       150      a3   
49999       1250    limousine               1996     manual       101  vectra   

       odometer_km  registration_month fuel_type       brand  \
0           150000                   3       lpg     peugeot   
1           150000                   6    benzin         bmw   
2            70000                   7    benzin  volkswagen   
3            70000                   6    benzin       smart   
4           150000                   7    benzin        ford   
...            ...                 ...       ...         ...   
49995       100000                   1    diesel        audi   
49996       150000                   5    benzin        opel   
49997         5000                  11    benzin        fiat   
49998        40000                  11    diesel        audi   
49999       150000                   1    benzin        opel   

      unrepaired_damage  ad_created  postal_code  last_seen  
0                    no    20160326        79588   20160406  
1                    no    20160404        71034   20160406  
2                    no    20160326        35394   20160406  
3                    no    20160312        33729   20160315  
4                    no    20160401        39218   20160401  
...                 ...         ...          ...        ...  
49995                no    20160327        82131   20160401  
49996                no    20160328        44807   20160402  
49997                no    20160402        73430   20160404  
49998                no    20160308        35683   20160405  
49999                no    20160313        45897   20160406  

[45778 rows x 16 columns]

### Analysis next steps

Finding the most common brand/model combinations

In [73]:
brand_model = autos.groupby(["brand", "model"]).size().reset_index(name="count")
brand_model_sorted = brand_model.sort_values("count", ascending=False)
brand_model_sorted.head(10)

brand     model  count
267     volkswagen      golf   3639
23             bmw       3er   2596
273     volkswagen      polo   1533
182           opel     corsa   1504
271     volkswagen    passat   1343
179           opel     astra   1318
13            audi        a4   1223
141  mercedes_benz  c_klasse   1134
24             bmw       5er   1123
144  mercedes_benz  e_klasse    952

From abowe tabel we can see that the three most common brand/model combinations are `Volkswagen GOlf`, `BMW 3ER` and `Volkswagen Polo`.

### Spliting the `odometer_km`

In [82]:
autos["odometer_km"].unique()

array([150000,  70000,  50000,  80000,  10000,  30000, 125000,  90000,
        20000,  60000,   5000,  40000, 100000])

In [91]:
import numpy as np

autos["mileage_group"] = pd.cut(autos["odometer_km"], bins=[0, 30000, 60000, 90000, 120000, np.inf], labels=["<30k", "30k-60k", "60k-90k", "90k-120k", ">120k"])
mileage_prices = autos.groupby("mileage_group")["price_usd"].mean().reset_index(name="avg_price")
mileage_prices

mileage_group     avg_price
0          <30k  15244.171224
1       30k-60k  13801.198552
2       60k-90k   9626.102924
3      90k-120k   8112.176703
4         >120k   4186.755683

In abowe table we can see that average price of second hand cars is lower for cars with high value of mileage  and is higer for cars with lower value of mileage.

#### How much cheaper are cars with damage than their non-damaged counterparts?

In [95]:
cars_with_demage = autos[autos["unrepaired_damage"] == "yes"]["price_usd"].mean()
cars_with_no_demage = autos[autos["unrepaired_damage"] == "no"]["price_usd"].mean()
price_diff = (cars_with_no_demage - cars_with_demage) / cars_with_no_demage * 100

print('Undamaged cars are, on average, {:.2f}% more expensive than damaged cars'.format(price_diff))

Undamaged cars are, on average, 66.92% more expensive than damaged cars


### Conclusion

In this analysis, we explored a dataset of used car listings on eBay Kleinanzeigen, cleaned the data, and performed some basic data exploration. We found that the dataset contained some inaccurate or unrealistic values, such as registration years outside the range of 1886-2016. So we clean the data to refers to year between 1886-2016 and delate columns with informations that weren't useful for our analysis.

We found that the top three car brands in the dataset:  BMW, Mercedes-Benz and Audi being more expensive and Renault, Fiat, and Opel being less expensive, while Volkswagen is in between. We also observed correlation between mileage and price, indicating that the most important thing for the price is the brand of the car, not the milesages. But on the other hand cars with less mileages value have higer average price than cars with high mileage values. In our analysis we also check that undemaged cars are more expensive than demage cars and the most popular brand/model combinations are Volkswagen GOlf, BMW 3ER and Volkswagen Polo. 

However, it's important to note that these observations are limited to the dataset we analyzed and may not necessarily be representative of the entire used car market.